src/memory/AmdFlash.cc

```diff
@@ -35,152 +41,125 @@ AmdFlash::AmdFlash(const std::string& name, const ValidatedChip& validatedChip,
-						ranges::fill(std::span{ramPtr, sectorSize}, 0xFF);
-					} else if (offset + sectorSize >= romSize) {
-						// partial overlap
-						auto last = romSize - offset;
-						auto missing = sectorSize - last;
-						const uint8_t* romPtr = &(*rom)[offset];
-						ranges::copy(std::span{romPtr, last}, ramPtr);
-						ranges::fill(std::span{&ramPtr[last], missing}, 0xFF);
-					} else {
-						// completely before end of rom
-						const uint8_t* romPtr = &(*rom)[offset];
-						ranges::copy(std::span{romPtr, sectorSize}, ramPtr);
-					}
-				}
+	size_t fileRemaining = file.is_open() ? file.getSize() : 0;
+	const std::span<const uint8_t> romData = rom ? std::span{*rom} : std::span<const uint8_t>{};
+	const std::span<uint8_t> ramData = ram.getWriteBackdoor();
+	for (size_t sector = 0, address = 0; const AmdFlash::Region& region : chip.geometry.regions) {
+		for (size_t regionSector = 0; regionSector < region.count; regionSector++, sector++, address += region.size) {
+			if (sectors[sector].modified) {
+				const size_t size = std::min<size_t>(region.size, fileRemaining);
+				if (size > 0) file.read(ramData.subspan(address, size));
+				ranges::fill(ramData.subspan(address + size, region.size - size), 0xFF);
+				fileRemaining -= size;
 			} else {
-				assert(rom); // must have rom constructor parameter
-				if ((offset + sectorSize) <= romSize) {
-					readAddress[sector] = &(*rom)[offset];
-				} else {
-					readAddress[sector] = nullptr;
-				}
+				const size_t size = std::min(address + region.size, romData.size()) - std::min(address, romData.size());
+				ranges::copy(romData.subspan(address, size), ramData.subspan(address, size));
+				ranges::fill(ramData.subspan(address + size, region.size - size), 0xFF);
 			}
-			offset += sectorSize;
 		}
 	}
-	assert(offset == size());
 
 	reset();
 }
 
-AmdFlash::~AmdFlash() = default;
+AmdFlash::~AmdFlash()
+{
+	if (schedulable.isPendingRT()) {
+		schedulable.cancelRT();
+		save();
+	}
+}
+
+void AmdFlash::markModified(size_t sector)
+{
+	assert(sector < sectors.size());
+	sectors[sector].modified = true;
+	if (!schedulable.isPendingRT()) {
+		schedulable.scheduleRT(5000000); // sync to disk after 5s
+	}
+}
+
+void AmdFlash::save()
+{
+	const auto& filename = config.getChildData("sramname");
+	try {
+		const std::string path = config.getFileContext().resolveCreate(filename);
+		File file(path, File::OpenMode::SAVE_PERSISTENT);
+		for (size_t sector = 0, address = 0; const AmdFlash::Region& region : chip.geometry.regions) {
+			for (size_t regionSector = 0; regionSector < region.count; regionSector++, sector++, address += region.size) {
+				if (sectors[sector].modified) {
+					file.write(std::span{ram}.subspan(address, region.size));
+				}
+			}
+		}
+		XmlOutputArchive out(path + ".meta");
+		out.serialize("sectors", sectors);
+	} catch (FileException& e) {
+		config.getCliComm().printWarning(
+			"Couldn't save SRAM ", filename, " (", e.getMessage(), ").");
+	}
```

Comment:
Because we've split the info over two files we've introduced some extra failure modes (maybe not likely, but possible):
* It's possible that only one of the two files gets written (e.g. some error during writing the second one).
* Then either that 2nd file will be missing (the bw-compat scenario). And it will be wrongly interpreted as-if all sectors have been written).
* Or the 2nd file is not written and a previous version of the file is kept. This may also lead to wrong conclusions.

Maybe this is an unlikely scenario? Should we do something about it? How much effort should we spend on this? Here are some possibilities:
* Think about the order in which we write these two files. Is one order better than the other (in case of a partial failure I mean)?
* Should we first remove the existing files before writing the new ones?
* We could write both to a temporary files (same directory, but different filename. e.g. with extra `.partial` suffix). And only when both writes were successful, rename the files (so possibly removing the old files).

--

An alternative could be to go back to a single file.

I proposed the two-file solution, assuming that the data-file would be a full-copy of the flash rom content. If that's not the case (when it can not be meaningfully interpreted without also knowing the meta-data), then I see less value in having it two separate files.

Since you're now using the serialization stuff it becomes relatively easy to embed binary data in the same file  (via `serialize_as_blob`). Maybe a bonus: this binary data will also be stored compressed (and base64 encoded).

--

I don't have a strong preference, I on;y wanted to point out some advantages/disadvantages of both approaches.

Reply:
> * Think about the order in which we write these two files. Is one order better than the other (in case of a partial failure I mean)?

Since the metadata file is meaningless without the data file, it’s probably best to write that one first, and if it fails, not write the data file.

Thinking about when this could fail, I think file creation permissions are the most likely rather than disk space. If you can create / modify the metadata file, then it is unlikely you can’t also create / modify the data file. Whereas if a data file already exists without metadata, it could be that it was created under different permission conditions and you could modify that but then fail to create the metadata.

> * Should we first remove the existing files before writing the new ones?

That can lead to unrecoverable data loss, so rather not…

> * We could write both to a temporary files (same directory, but different filename. e.g. with extra `.partial` suffix). And only when both writes were successful, rename the files (so possibly removing the old files).

I think that’s a bit over the top, that’s not done anywhere else in openMSX afaik even if there could very well be situations where you could in theory e.g. write an incomplete file when running out of disk space.

A bad alternative that also came to mind: write one file per sector :).

> An alternative could be to go back to a single file.
>
> I proposed the two-file solution, assuming that the data-file would be a full-copy of the flash rom content. If that's not the case (when it can not be meaningfully interpreted without also knowing the meta-data), then I see less value in having it two separate files.

The reason why I went for the sparse implementation immediately is because it integrated a bit more neatly with the existing method, back compat was easy to do, and it’s also more space-efficient with the savegame sectors use case in mind. But, I’m not necessarily against storing a full copy.

If it’d store a full copy, the data in the non-modified sectors would be meaningless.

  * If it stored ROM data (straight dump of the tracked RAM) it could be outdated compared to the ROM passed in. But it’s kinda cool that’d then be a usable .rom actually.
  * If it filled those unmodified parts with FFH or something, fine I guess, feels a bit unsatisfying tho.

> Since you're now using the serialization stuff it becomes relatively easy to embed binary data in the same file (via `serialize_as_blob`). Maybe a bonus: this binary data will also be stored compressed (and base64 encoded).

Downside is that this does introduce a new file format, so it’d mean a more elaborate back compat code path, a new file extension, etc. Also, maybe I needn’t worry about this, but wouldn’t it be slower due to the encoding and compression? These can be fairly large (8 MB) files… Then again, they won’t change often, so probably ok.

Something also to consider here; a FlashROM can persist more data than just the ROM data. Some Flash ROMs like the M29W640G and S29GL064S have a 256-byte “extended memory block” or “secure silicon region”, currently it’s not implemented, but let’s say it would be sometime in the future, how do we store that? If we have separate files I don’t think it makes sense to append to the plain data, I would want to either write it as a separate file or include it in the metadata.

Anyway not really sure what’s best here either. Let me push some updates and then see if we can make up our minds. The serialisation details are relatively easy to change.

Reply:
I’ve had some more time to think about it, and I think a new single-file `.flash` format is the way to go. It’s just much more easily extensible this way. Future extensions of FlashROM support may include persisting the extended memory block, and also the S29GL064S supports ECC information, etc. It would make no sense to save these as separate files, so why is the main Flash ROM data special. Just extra hassle.